multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapCacheManager.java

```diff
@@ -16,21 +31,114 @@
+            .<ScopedState, EmbeddedMultimapConfiguration>getCache(GlobalConfigurationManager.CONFIG_STATE_CACHE_NAME)
+            .getAdvancedCache();
    }
 
    @Override
    public Configuration defineConfiguration(String name, Configuration configuration) {
-      return cacheManager.defineConfiguration(name, configuration);
+      Configuration registered = cacheManager.defineConfiguration(name, configuration);
+      boolean defined = false;
+      MultimapCacheManagerConfiguration cmc = configuration.module(MultimapCacheManagerConfiguration.class);
+      if (cmc != null) {
+         EmbeddedMultimapConfiguration mc = cmc.multimaps().get(name);
+         if (mc != null) {
+            CompletionStages.join(defineConfigurationAsync(mc));
+            defined = true;
+         }
+      }
+
+      if (!defined) {
+         EmbeddedMultimapConfiguration mc = configuration.module(EmbeddedMultimapConfiguration.class);
+         if (mc != null && mc.name().equals(name)) {
+            CompletionStages.join(defineConfigurationAsync(mc));
+         }
+      }
+      return registered;
+   }
+
+   @Override
+   public CompletionStage<Boolean> defineConfiguration(Configuration cacheConfiguration, EmbeddedMultimapConfiguration multimapConfiguration) {
+      cacheManager.defineConfiguration(multimapConfiguration.name(), cacheConfiguration);
+      return defineConfiguration(multimapConfiguration);
+   }
+
+   @Override
+   public CompletionStage<Boolean> defineConfiguration(EmbeddedMultimapConfiguration configuration) {
+      return defineConfigurationAsync(configuration);
+   }
+
+   private CompletionStage<Boolean> defineConfigurationAsync(EmbeddedMultimapConfiguration configuration) {
+      final String name = configuration.name();
+      return checkGlobalConfiguration(name)
+            .thenCompose(existent -> {
+               if (existent != null) {
+                  // Already statically defined.
+                  return CompletableFutures.completedFalse();
+               }
+
+               return stateCache.putIfAbsentAsync(stateKey(name), configuration)
+                     .thenApply(Objects::isNull);
+            });
+   }
+
+   private CompletionStage<EmbeddedMultimapConfiguration> checkGlobalConfiguration(String name) {
+      for (EmbeddedMultimapConfiguration configuration : preConfigured.values()) {
+         if (configuration.name().equals(name)) {
+            return stateCache.putIfAbsentAsync(stateKey(name), configuration)
+                  .thenApply(c -> c == null ? configuration : c);
+         }
+      }
+      return CompletableFutures.completedNull();
+   }
+
+   private static ScopedState stateKey(String name) {
+      return new ScopedState(MULTIMAP_SCOPE, name);
+   }
+
+   @Override
+   public MultimapCache<K, V> get(String name) {
+      CompletionStage<MultimapCache<K, V>> cs = caches.computeIfAbsent(name, this::createMultimapCache);
+      return CompletionStages.join(cs);
```

Comment:
Are these joins an issue? They happen on the `get` and on the `defineConfiguration` method.

Reply:
not for now, but definitely needs to be double checked

Reply:
If these methods are going to be called by any of our threads, yes it is an issue.